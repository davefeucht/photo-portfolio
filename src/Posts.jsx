import React from 'react';
import ReactDOM from 'react-dom';
import axios from 'axios';
import Post from './Post.jsx';

class Posts extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      posts: [],
      showAllPosts: true,
      categoryData: {},
      singlePostToShow: 0,
      errorMsg: ""
    }
  }

  _getPosts() {
    axios.get(`http://` + this.props.site + `/wp-json/wp/v2/posts?categories=` + this.props.category)
    .then(res => {
      const posts = res.data;
      this.setState({ posts });
    }, error => {
      const errorMsg = 'Did not work: ' + (error.response ? error.response : error);
      this.setState({ errorMsg });
    });
  }

  _getPostsCategory() {
    axios.get(`http://` + this.props.site + `/wp-json/wp/v2/categories/` + this.props.category)
    .then(res => {
      const categoryData = res.data;
      this.setState({ categoryData });
    }, error => {
      const errorMsg = 'Did not work: ' + (error.response ? error.response : error);
      this.setState({ errorMsg });
    });
  }
  
  _showAllPosts() {
    const showAllPosts = true;
    const singlePostToShow = 0;
    this.setState({ showAllPosts });
    this.setState({ singlePostToShow }); 
  }

  componentWillMount() {
    this._getPosts();
    this._getPostsCategory();
  }

  render() {
    let postList = [];
    let categoryHeader = "";
    if(this.state.showAllPosts) {
      postList = this.state.posts.map(post =>
        { return ( <Post key={post.id.toString()} category={this.props.category} categoryName={this.state.categoryData.name} title={post.title.rendered} id={post.id} image={post.featured_media} context="thumbnail" site={this.props.site} /> ); }
      );
      categoryHeader = <h2 className="category-title">{this.state.categoryData.name}</h2>;
    }
    else {
      this.state.posts.forEach(post => {
        if(post.id == this.state.singlePostToShow) {
          postList[0] = <Post key={this.state.singlePostToShow.toString()} id={this.state.singlePostToShow} category={this.props.category} categoryName={this.state.categoryData.name} title={post.title.rendered} image={post.featured_media} context="full-image" site={this.props.site} />;
        }
      });
    }
    
    return (
      <div>
        {postList} 
        <div>{this.state.errorMsg}</div>
      </div>
    );
  }
}

export default Posts;
