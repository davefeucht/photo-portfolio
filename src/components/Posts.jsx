/*****************
* Posts component displays a list of Posts
*****************/

import React from "react";
import axios from "axios";
import Post from "./Post.jsx";
import PostThumbnail from "./PostThumbnail.jsx";
import PostList from "./styledComponents/PostList.jsx";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {bindActionCreators} from "redux";
import {setCategoryPosts, setCategoryData, setShowAllPosts, setSinglePostToShow} from "../actions/actions.js";

class Posts extends React.Component {
  static propTypes = {
    postCategory: PropTypes.number,
    siteUrl: PropTypes.string,
    categoryData: PropTypes.object,
    categoryPosts: PropTypes.array,
    showAllPosts: PropTypes.bool,
    singlePostToShow: PropTypes.number
  };

  //Function to get all posts for a given category
  _getPosts() {
    const getPostsURI = `https://${this.props.siteUrl}/wp-json/wp/v2/posts?categories=${this.props.postCategory}`;
    axios.get(getPostsURI)
      .then(res => {
        const posts = res.data;
        this.props.setCategoryPosts(posts);
      });
  }

  //Function to get data about the category in which these posts are found
  _getPostsCategory() {
    const getPostsCategoryURI = `https://${this.props.siteUrl}/wp-json/wp/v2/categories/${this.props.postCategory}`; 
    axios.get(getPostsCategoryURI)
      .then(res => {
        const categoryData = res.data;
        this.props.setCategoryData(categoryData);
      });
  }
  
  //Function to show a single post when clicked
  _showSinglePost(postId) {
    const showAllPosts = false;
    const name = "";
    this.props.setShowAllPosts(showAllPosts);
    this.props.setSinglePostToShow({postId, name});
  }

  //Function to show all posts in the category
  _showAllPosts(postId) {
    const showAllPosts = true;
    const name = "";
    this.props.setShowAllPosts(showAllPosts);
    this.props.setSinglePostToShow({postId, name}); 
  }

  //Function to show all the categories - onClick of 'Back To Categories' text
  _showAllCategories() {
    this.props.clickCategory();
  }

  //When the component is about to mount, get the posts for the category and get the category info
  componentWillMount() {
    this._getPosts();
    this._getPostsCategory();
  }

  render() {
    let postList = [];

    //If we are showing all posts, then map the list of posts to a list of PostThumbnail components
    if(this.props.showAllPosts) {
      postList = this.props.categoryPosts.map(post =>
      { return ( <PostThumbnail key={post.id.toString()} id={post.id} image={post.featured_media} site={this.props.siteUrl} clickImage={this._showSinglePost.bind(this)}/> ); }
      );
    }
    
    //Otherwise, display the single post which matches the 'singlePostToShow'
    else {
      this.props.categoryPosts.forEach(post => {
        if(post.id == this.props.singlePostToShow) {
          postList[0] = <Post key={this.props.singlePostToShow.toString()} id={this.props.singlePostToShow} category={this.props.postCategory} categoryName={this.props.categoryData.name} title={post.title.rendered} image={post.featured_media} context="full-image" site={this.props.siteUrl} clickImage={this._showAllPosts.bind(this)} />;
        }
      });
    }
    
    return (
      <PostList>
        {postList} 
      </PostList>
    );
  }
}

const mapStateToProps = state => {
  return {
    postCategory: state.visibilityFilter.singleCategoryToShow.categoryId, 
    siteUrl: state.applicationState.siteUrl,
    categoryData: state.applicationState.currentCategoryData,
    categoryPosts: state.applicationState.currentCategoryPosts,
    showAllPosts: state.visibilityFilter.showAllPosts,
    singlePostToShow: state.visibilityFilter.singlePostToShow.postId
  };
}

const mapDispatchToProps = dispatch => {
  return {
    setCategoryPosts: bindActionCreators(setCategoryPosts, dispatch),
    setCategoryData: bindActionCreators(setCategoryData, dispatch),
    setShowAllPosts: bindActionCreators(setShowAllPosts, dispatch),
    setSinglePostToShow: bindActionCreators(setSinglePostToShow, dispatch)
  }; 
}

export default connect(mapStateToProps, mapDispatchToProps)(Posts);