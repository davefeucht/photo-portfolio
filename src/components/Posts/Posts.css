.posts {
  height: 100%;
  grid-area: category-content;
  display: grid;
  grid-template-columns: 20% 20% 20% 20% 20%;
  grid-template-rows: 1fr 1fr; 
}