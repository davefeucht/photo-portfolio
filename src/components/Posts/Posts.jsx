/*****************
* Posts component displays a list of Posts
*****************/

import React from 'react';
import { observer } from 'mobx-react';
import Post from '../Post/Post.jsx';
import PostThumbnail from '../PostThumbnail/PostThumbnail.jsx';
import './Posts.css';

const Posts = observer(({ stateStore, categoryId, categoryName, api }) => {

  //Function to show a single post when clicked
  const _showSinglePost = (postId) => {
    const showAllPosts = false;
    const name = "";
    //props.setShowAllPosts(showAllPosts);
    //props.setSinglePostToShow({postId, name});
  }

  //Function to show all posts in the category
  const _showAllPosts = (postId) => {
    const showAllPosts = true;
    const name = "";
    //props.setShowAllPosts(showAllPosts);
    //props.setSinglePostToShow({postId, name}); 
  }

  //Function to show all the categories - onClick of 'Back To Categories' text
  const _showAllCategories = () => {
    //props.clickCategory();
  }

  //When the component is about to mount, get the posts for the category and get the category info
  api.getPosts(categoryId);
  api.getCategoryInfo(categoryId);

  let postList = [];

  //If we are showing all posts, then map the list of posts to a list of PostThumbnail components
  if(stateStore.visibilityFlags.showAllPosts) {
    postList = stateStore.currentCategoryPosts.map((post, index) => { 
      return ( <PostThumbnail key={post.id.toString()} id={post.id} index={index} image={post.featured_media} clickImage={_showSinglePost.bind(this)}/> ); 
    });
  }
    
  //Otherwise, display the single post which matches the 'singlePostToShow'
  else {
    stateStore.currentCategoryPosts.forEach(post => {
      if(post.id == stateStore.visiblePost.postId) {
        postList[0] = <Post key={stateStore.visiblePost.postId.toString()} id={stateStore.visiblePost.postId} category={categoryId} categoryName={categoryName} title={post.title.rendered} image={post.featured_media} context="full-image" clickImage={_showAllPosts.bind(this)} />;
      }
    });
  }
    
  return (
    <div className="posts">
      <div>
        {postList} 
      </div>
    </div>
  );
});

Posts.displayName = 'Posts';

export default Posts;