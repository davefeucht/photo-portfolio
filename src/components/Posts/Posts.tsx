/** ***************
* Posts component displays a list of Posts
**************** */

import './Posts.css';

import { observer } from 'mobx-react';
import * as React from 'react';
import { useEffect, useState } from 'react';

import { Store } from '../../utils/types';
import PaginationNavigation from '../PaginationNavigation/PaginationNavigation';
import PostThumbnail from '../PostThumbnail/PostThumbnail';

interface PostsProps {
    stateStore: Store
}

const Posts: React.FC<PostsProps> = ({ stateStore }) => {
    const [currentPageIndex, setCurrentPageIndex] = useState(1);

    const startIndex = (currentPageIndex - 1) * stateStore.maxItemsPerPage;
    const endIndex = startIndex + stateStore.maxItemsPerPage;

    const setRows = () => {
        const numberOfColumns = parseInt(document.body.style.getPropertyValue('--number-of-columns'));
        const numberOfPosts = stateStore.maxItemsPerPage;
        document.body.style.setProperty('--number-of-rows', `${numberOfPosts / numberOfColumns}`);
    };

    useEffect(() => {
        setRows();
    }, [stateStore.screenInfo.width, stateStore.screenInfo.height]);

    return (
        <div className="posts">
            {stateStore.currentCategoryPosts.slice(startIndex, endIndex)
                .map((post, index) => {
                    return (<PostThumbnail key={post.id.toString()} stateStore={stateStore} id={post.id} index={index + startIndex} />);
                })}
            <PaginationNavigation totalPages={stateStore.currentCategoryPosts.length / stateStore.maxItemsPerPage} currentPageIndex={currentPageIndex} navigationFunction={content => { setCurrentPageIndex(content); }} />
        </div>
    );
};

Posts.displayName = 'Posts';

export default observer(Posts);
