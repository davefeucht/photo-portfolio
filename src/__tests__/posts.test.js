import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import Posts from '../components/Posts/Posts';
import stateStore from '../StateStore/store';

const posts = [
    {
        id: 887,
        featured_media: 142,
        categories: [220],
        title: {
            rendered: 'Some post'
        }
    },
    {
        id: 889,
        featured_media: 145,
        categories: [220],
        title: {
            rendered: 'Some other post'
        }
    }
];

test('Posts displays', () => {
  const store = new stateStore();
  let component;
  let tree;
  store.setCategoryPosts(posts);
  component = renderer.create(
      <MemoryRouter>
          <Posts stateStore={store}></Posts>
      </MemoryRouter>
  );
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
