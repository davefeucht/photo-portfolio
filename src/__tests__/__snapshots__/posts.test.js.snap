// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts displays 1`] = `
<div
  className="posts"
>
  <a
    href="/post/887"
    onClick={[Function]}
  >
    <div
      className="post-thumbnail"
      style={
        {
          "backgroundImage": "url()",
        }
      }
    />
  </a>
  <a
    href="/post/889"
    onClick={[Function]}
  >
    <div
      className="post-thumbnail"
      style={
        {
          "backgroundImage": "url()",
        }
      }
    />
  </a>
</div>
`;
